src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpContext.ReadWrite.cs

```diff
@@ -0,0 +1,431 @@
+                            }
+                            else
+                            {
+                                // Flush of zero bytes
+                                await FlushToIISAsync();
+                            }
+                        }
+                        finally
+                        {
+                            // Always Advance the data pointer to the end of the buffer.
+                            Output.Reader.AdvanceTo(buffer.End);
+                        }
+                    }
+
+                    // Check if there was an upgrade. If there is, we will replace the request and response bodies with
+                    // two seperate loops. These will still be using the same Input and Output pipes here.
+                    if (_upgradeTcs?.TrySetResult(null) == true)
+                    {
+                        await StartBidirectionalStream();
+
+                        // Input and Output will be closed in StartBidirectionalStream.
+                        // We can return at this point.
+                        return;
+                    }
+
+                    // Now we handle the read. 
+                    var memory = Input.Writer.GetMemory();
+                    _inputHandle = memory.Retain(true);
+
+                    try
+                    {
+                        // Lock around invoking ReadFromIISAsync as we don't want to call CancelIo
+                        // when calling read
+                        var read = await ReadFromIISAsync(memory.Length);
+
+                        // read value of 0 == done reading
+                        // read value of -1 == read cancelled, continue trying to read
+                        // but don't advance the pipe
+                        if (read == 0)
+                        {
+                            _doneReading = true;
+                        }
+                        else if (read != -1)
+                        {
+                            Input.Writer.Advance(read);
+                        }
+                    }
+                    finally
+                    {
+                        // Always commit any changes to the Input pipe
+                        Input.Writer.Commit();
+                        _inputHandle.Dispose();
+                    }
+
+                    // Flush the read data for the Input Pipe writer
+                    var flushResult = await Input.Writer.FlushAsync();
+
+                    // If the pipe was closed, we are done reading, 
+                    if (flushResult.IsCompleted || flushResult.IsCancelled)
+                    {
+                        _doneReading = true;
+                    }
+                }
+
+                // Complete the input writer as we are done reading the request body.
+                Input.Writer.Complete();
+            }
+            catch (Exception ex)
+            {
+                Output.Reader.Complete(ex);
```

Comment:
The reader exception doesn't matter as much. The writer exception is about triggering errors in ReadAsync. This will flow to user code so we should make sure that we understand what those can look like.